docs/development/etcdoperatortask-design.md

```diff
@@ -0,0 +1,501 @@
+
+---
+
+## Custom Resource API Design
+
+The `EtcdOperatorTask` CRD is defined under the `v1alpha1` API version. It is subject to change and will follow the [Kubernetes Deprecation Policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/).
+
+**Key design decisions:**
+
+- The `spec` is immutable (`kubebuilder:validation:Immutable`) to ensure task intent cannot be changed after creation.
+- The `config` field uses `runtime.RawExtension` to flexibly support task-specific parameters, validated via admission webhook.
+- `ownerEtcdReference` under the spec field is renamed to `etcdReference` for clarity.
+- The `LastOperation` struct now uses a `Description` field instead of `Name` and `Reason` for better expressiveness.
+
+---
+
+### Go API Definition
+
+<details>
+<summary>Show EtcdOperatorTask CRD Go Definition</summary>
+
+```go
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=etcdoperatortasks,shortName=eot;eots,scope=Namespaced
+// +kubebuilder:subresource:status
+type EtcdOperatorTask struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec   EtcdOperatorTaskSpec   `json:"spec"`
+    
+    Status EtcdOperatorTaskStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:validation:XPreserveUnknownFields
+// +kubebuilder:validation:Immutable
+type EtcdOperatorTaskSpec struct {
+    // +required
+    Type v1alpha1.EtcdOperatorTaskType `json:"type"`
+
+    // Config is task-specific key/value parameters.
+    // +required
+    Config runtime.RawExtension `json:"config,omitempty"` 
+
+    // TTLSecondsAfterFinished controls how long the status+pod stays around.
+    // +optional
+    // +kubebuilder:validation:XValidation:message="TTLSecondsAfterFinished must be greater than 0",rule="!has(self) || self > 0"
+    TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+
+    // +optional
+    EtcdReference types.NamespacedName `json:"etcdReference,omitempty"`
+
+}
+
+```
+</details>
+
+> **Best Practice:**
+> Use `runtime.RawExtension` for the `config` field to allow each task type to define its own schema. Validate this field with an admission webhook for type safety and user feedback.
+
+---
+
+### Status Subresource
+
+The `status` field tracks the progress and outcome of each task, including state transitions, errors, and operation history.
+
+<details>
+<summary>Show EtcdOperatorTaskStatus Go Definition</summary>
+
+```go
+// EtcdOperatorTaskStatus is the status for a EtcdOperatorTask resource.
+type EtcdOperatorTaskStatus struct {
+  // State is the last known state of the task.
+  State TaskState `json:"state"`
+  // Time at which the task has moved from "pending" state to any other state.
+  // +optional
+  InitiatedAt *metav1.Time `json:"initiatedAt"`
```

Comment:
If `Status.State` is `Rejected`, then would `initiatedAt` still record the time at which it moved from `Pending` to `Rejected`? If yes, then the name seems non-intuitive, since the task was not really initiated, since it was rejected.

Reply:
As we discussed out of band discussion, The `status.initiatedAt` field will be set when the task transitions from `Pending` to `InProgress`.